### PR TITLE
GraphQL Add-On Updates

### PR DESCRIPTION
**Pull request by ricekot**
<details>
<summary>Fix variable names clash</summary>
If we have a schema:

```graphql
type Query {
    field1: type1
    field2: type2
}

type type1 {
    name (id:ID): String
}

type type2 {
    name (id: ID): String
}
```
Here's what the GraphQL add-on currently generates:

```graphql
query ($name_id: ID, $name_id: ID) {
    field1 {
	name (id: $name_id)
    }
    field2 {
	name (id: $name_id)
    }
}
```
As we can observe, there is a clash in the names of the variables.

After this fix, the add-on will generate:

```graphql
query ($field2_name_id: ID, $field1_name_id: ID) {
    field1 {
	name (id: $field1_name_id)
    }
    field2 {
	name (id: $field2_name_id)
    }
}
```
...as it should.
</details>

- Add Support for Inline Arguments Injection
	- Add helper class InlineInjector for inline arguments injection using script input vectors.
	- Add unit tests for this class.
- Fix NPEs during Spidering
	- Add null check for the Content-Type Header in MessageValidator
	- Add a unit test for the null check
- Add Input Vectors Script
	- Bundle it with the add-on
- Add Help Page for the script
- As discussed on IRC, 
	- Add null check for import message in Introspection
	- Reduce Log Level for the Requestor

Signed-off-by: ricekot <ricekot@gmail.com>

### CODE REVIEW

**Review by kingthorin — COMMENTED** [review]
Seems okay to me.

**Review by kingthorin — APPROVED** [review]
Seems good to me, thanks

**Comment by ricekot** [comment]
This pull request is ready for a review! :)

File: addOns/graphql/src/main/java/org/zaproxy/addon/graphql/ExtensionGraphQl.java [hunk 1]
```diff
@@ -165,6 +181,33 @@ private void stopParserThreads() {
...
+        if (extScript != null) {
+            ScriptType variantType =
+                    extScript.getScriptType(ExtensionActiveScan.SCRIPT_TYPE_VARIANT);
+            ScriptEngineWrapper engine =
+                    extScript.getEngineWrapper(
+                            Constant.messages.getString("graphql.script.engine"));
+            if (variantType != null && engine != null) {
+                String scriptName = Constant.messages.getString("graphql.script.name");
+                File scriptPath =
+                        new File(Constant.getZapHome(), "scripts/scripts/variant/" + scriptName);
```
[the remark this answers is not in the document]
Comment (thc202) [reply addOns/graphql/src/main/java/org/zaproxy/addon/graphql/ExtensionGraphQl.java:196]: Core provides several constants, e.g.:
```java
Paths.get(Constant.getZapHome(), ExtensionScript.SCRIPTS_DIR, ExtensionScript.SCRIPTS_DIR, ExtensionActiveScan.SCRIPT_TYPE_VARIANT, scriptName).toFile()
```

File: addOns/graphql/src/main/java/org/zaproxy/addon/graphql/ExtensionGraphQl.java [hunk 2]
```diff
@@ -60,6 +65,7 @@
     private static final int ARG_IMPORT_FILE_IDX = 0;
     private static final int ARG_IMPORT_URL_IDX = 1;
     private static final int ARG_END_URL_IDX = 2;
+    private static final String FIELD_ENGINE = "scripts.dialog.script.label.engine";
```
Comment (thc202) [inline addOns/graphql/src/main/java/org/zaproxy/addon/graphql/ExtensionGraphQl.java:68]: Does not seem to be used?

File: addOns/graphql/src/main/java/org/zaproxy/addon/graphql/ExtensionGraphQl.java [hunk 3]
```diff
@@ -165,6 +181,33 @@ private void stopParserThreads() {
         parserThreads.clear();
     }
 
+    private void addScript() throws IOException {
+        ExtensionScript extScript =
+                Control.getSingleton().getExtensionLoader().getExtension(ExtensionScript.class);
+        if (extScript != null) {
```
Comment (thc202) [inline addOns/graphql/src/main/java/org/zaproxy/addon/graphql/ExtensionGraphQl.java:187]: `&& extScript.getScript(scriptName) == null`

File: addOns/graphql/src/main/java/org/zaproxy/addon/graphql/ExtensionGraphQl.java [hunk 4]
```diff
@@ -165,6 +181,33 @@ private void stopParserThreads() {
...
+            ScriptType variantType =
+                    extScript.getScriptType(ExtensionActiveScan.SCRIPT_TYPE_VARIANT);
+            ScriptEngineWrapper engine =
+                    extScript.getEngineWrapper(
+                            Constant.messages.getString("graphql.script.engine"));
+            if (variantType != null && engine != null) {
+                String scriptName = Constant.messages.getString("graphql.script.name");
```
Comment (thc202) [inline addOns/graphql/src/main/java/org/zaproxy/addon/graphql/ExtensionGraphQl.java:194]: Might be better to hardcode, otherwise the user might end up with several scripts (e.g. when changing language).

File: addOns/graphql/src/main/java/org/zaproxy/addon/graphql/ExtensionGraphQl.java [hunk 5]
```diff
@@ -165,6 +181,33 @@ private void stopParserThreads() {
...
 
+    private void addScript() throws IOException {
+        ExtensionScript extScript =
+                Control.getSingleton().getExtensionLoader().getExtension(ExtensionScript.class);
+        if (extScript != null) {
+            ScriptType variantType =
+                    extScript.getScriptType(ExtensionActiveScan.SCRIPT_TYPE_VARIANT);
+            ScriptEngineWrapper engine =
+                    extScript.getEngineWrapper(
+                            Constant.messages.getString("graphql.script.engine"));
```
Comment (thc202) [inline addOns/graphql/src/main/java/org/zaproxy/addon/graphql/ExtensionGraphQl.java:192]: The engine name can be hardcoded.

File: addOns/graphql/src/main/java/org/zaproxy/addon/graphql/GraphQlParser.java [hunk 6]
```diff
@@ -64,21 +64,22 @@ public void introspect() throws IOException {
...
-                                    new TypeToken<Map<String, Object>>() {}.getType());
-            @SuppressWarnings("unchecked")
-            Document schema =
-                    new IntrospectionResultToSchema()
-                            .createSchemaDefinition((Map<String, Object>) result.get("data"));
-            String schemaSdl = new SchemaPrinter().print(schema);
-            parse(schemaSdl);
-        } catch (JsonSyntaxException e) {
-            throw new IOException("The response was not valid JSON.");
+        if (importMessage != null) {
```
Comment (thc202) [inline addOns/graphql/src/main/java/org/zaproxy/addon/graphql/GraphQlParser.java:67]: Could check it's null and return to avoid nesting.

File: addOns/graphql/src/main/java/org/zaproxy/addon/graphql/InlineInjector.java [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+                .map(defn -> (OperationDefinition) defn)
+                .map(OperationDefinition::getOperation)
+                .map(Object::toString)
+                .map(s -> s.toLowerCase(Locale.ROOT))
+                .distinct()
+                .sorted()
+                .collect(Collectors.joining(", "));
+    }
+
+    public Boolean validateQuery(String query) {
```
Comment (thc202) [inline addOns/graphql/src/main/java/org/zaproxy/addon/graphql/InlineInjector.java:315]: `boolean`

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thanks!

**Review by psiinon — APPROVED** [review]
